[Block Library - Post Featured Image]: Add basic dimension controls

ntsekouras opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Resolves: https://github.com/WordPress/gutenberg/issues/27620

This PR enables basic dimension controls (`width`, `height`) and allow `object-fit` properties to be configured. 
I believe landing at least a basic handling for now like this would be okay for now and gives value, in order to make it to 5.8.

## Notes
1. The `scale` control from the design does not exists now, so I've added some adhoc styles. We might improve those styles or consider adding a new component to match the design.
2. In order to set some `object-fit` properties we need to have set both `width` and `height` (see [comment](https://github.com/WordPress/gutenberg/issues/27620#issuecomment-833336369)).
3. In `scale` control I've implemented `toggle/unset` functionality. We can keep this or add a `reset/clear` button.
4. `height` in percentage (`%`) feels a bit unintuitive IMO - any suggestions?

## Testing instructions
1. Insert `Post Featured Image` block (single or in a `Query` block) and set `width`, `height` and `scale`. What feels natural to me is having `width` in `%` and `height` in `px`, but this is just subjective.
2. Observe editor and front-end changes.

ntsekouras committed:
[Block Library - Post Featured Image]: Add basic dimension controls

jasmussen commented:
What an incredibly strong first offering. To me. this PR feels like it validates the idea of the object-fit size controls appearing after you've set both dimensions. 

I'm seeing some weirdness with the sizing, though:

![featured](https://user-images.githubusercontent.com/1204802/117621141-91c59d00-b171-11eb-8703-72c97a2c8c54.gif)

Percent here doesn't seem to be working correctly, is it? Maybe it is. But 

vw seems to be faring better:

![take 2](https://user-images.githubusercontent.com/1204802/117621811-452e9180-b172-11eb-887f-138cfc41a12c.gif)

For some reason the scale values don't appear to work at small dimensions:

![contain](https://user-images.githubusercontent.com/1204802/117621866-54154400-b172-11eb-9bf7-ea7d0707b25b.gif)

A few small things:

- Default to PX unit — can we do that? If not, we might want to disable % by default.
- There should be a default scale setting. `fill` is the default value for object-fit, so we could use that. But `cover` feels more useful — can we set that by default?
- The new segmented control-style for this button group will need a little love. I'd love to help push some when you're ready for it.
- Width and Height are both `label`s. Should `Scale` be a label for the entire segmented control?

Excellent work! Excited to see this land.

jameskoster commented:
> The new segmented control-style for this button group will need a little love.

+1. Technical users will understand these terms, otherwise they have little meaning. 

Sharing for potential inspiration there – YouTube playback quality options are listed as "1080p, 720p, 480p", etc on desktop. But I recently noticed that the mobile apps have a different take on that:

<img src="https://user-images.githubusercontent.com/846565/117628454-1532be80-b171-11eb-978d-07ce8da75bf7.PNG" width="375" />

Maybe we can do something similar?

jasmussen commented:
To clarify I actually like the segmented control for this one. I was referring more to making it match [the mockup](https://github.com/WordPress/gutenberg/issues/27620#issuecomment-833336369).

That said, a dropdown would afford additional help text, that does make sense too. 

But the thing I like the most about the segmented control is that its effect is so instant — you just click through the three options and the results are immediately visually evident, teaching you what each option does. Combined with the fact that there is a default, I think this would be a good place to start — that would also let us make headway on the control itself, which is needed elsewhere also.

One thing we could do is to just add some help text below, to clarify, and to change the "Fill" to "Stretch":

<img width="1440" alt="Featured Image i2" src="https://user-images.githubusercontent.com/1204802/117630467-811a2480-b17b-11eb-996c-bf44e801548b.png">

jameskoster commented:
> the thing I like the most about the segmented control is that its effect is so instant — you just click through the three options and the results are immediately visually evident

Totally! I wasn't suggesting to change that. My point was mostly around the language. A label like "Contain" is going to be meaningless to the vast majority of people. You're correct that the canvas gives visual feedback – but only in the current context – it doesn't actually teach you what "Contain" means holistically. A few more words of description (or even just better labels, I like the "Stretch" example) might help make these more technical controls more digestible, and better assist visually impaired users?

Sorry for de-railing this PR, we can discuss this in a separate issue :D

jasmussen commented:
Not derailing at all, in fact very much on point. Stretch is one such step. But I didn't find a good single-word term to replace "contain". Something that encompasses "don't crop and don't stretch".

jameskoster commented:
Yes I don't think a single word will really cut it. 

We have this area control in the template part creation flow:

<img src="https://user-images.githubusercontent.com/5323259/114462542-7b183e80-9be3-11eb-9bfa-69f084801731.png">

Maybe we could use something like that, with similar descriptions to those found at https://www.w3schools.com/css/css3_object-fit.asp ?

Or, could we display a thumbnail preview, kind of like the style picker?

<img width="279" alt="Screenshot 2021-05-10 at 10 27 15" src="https://user-images.githubusercontent.com/846565/117638035-58ddf600-b17a-11eb-8d81-49ef95c596e8.png">

That would totally eliminate the mystery and the need to click through each option to visualise its effects. A tooltip could reveal the detailed description 🤔

jasmussen commented:
The thumbnail preview might work. But we'd need a great image where the differential between the three styles is clear.

ntsekouras committed:
adress feedback

ntsekouras commented:
I addressed some of the feedback:
1. `px` as starting unit
2. default to `cover`
3. change wording to `stretch`

>Width and Height are both labels. Should Scale be a label for the entire segmented control?

I tried finding a nice way to do this but found some problems with the [label behavior](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label). 
>When a <label> is clicked or tapped and it is associated with a form control, the resulting click event is also raised for the associated control.

>For some reason the scale values don't appear to work at small dimensions:

I think this is resolved now. I added `width:100%` to the image as our width control is affecting the parent `figure` element.

> for this button group will need a little love

Feel free to push any enhancements you like 😄

jasmussen commented:
I'm tinkering with some things and encountered this situation:

![featured](https://user-images.githubusercontent.com/1204802/117667571-4a590400-b1a5-11eb-9bc4-841274841a1d.gif)

Observe how if you add only a height, the image immediately gets super stretched.

If, however, you only add a width, the image looks fine and non-stretched.

This is because we have the following 100% width rule:

<img width="379" alt="Screenshot 2021-05-10 at 15 31 40" src="https://user-images.githubusercontent.com/1204802/117667716-74aac180-b1a5-11eb-8b40-3cef1a852f83.png">

In other words, if you add only a height but no width, you actually do have a width — 100%. Which would also benefit from the "cover" scaling property.

Can we try a change so that the cover/contain/stretch control also becomes available if you add just a height? I.e. show the control if there's a height, or a height and a width.

jasmussen commented:
Another question, I'm looking at the segmented control:

<img width="457" alt="Screenshot 2021-05-10 at 15 39 40" src="https://user-images.githubusercontent.com/1204802/117668171-f864ae00-b1a5-11eb-8454-c88c94778127.png">

It looks like a one-off for the featured control here for now, and that might be a fine way to dip our toes in? Alternately, should we already now start to extract this one as its own control that can be used in other places? We'll need it for the nav block also https://github.com/WordPress/gutenberg/issues/30370

Otherwise let me know and I'll just style according to the `block-library-post-featured-image-scale-controls` class.

ntsekouras committed:
add scale if height is set

ntsekouras commented:
>It looks like a one-off for the featured control here for now, and that might be a fine way to dip our toes in? Alternately, should we already now start to extract this one as its own control that can be used in other places? 

That I believe is more of a design folks decision. IMO this functionality will be much needed and 5.8 is really close 😄  Either way I'm happy to help creating the new control if needed now.

--edit:
>Otherwise let me know and I'll just style according to the block-library-post-featured-image-scale-controls class.

If we are okay to style this one, your help would be great for sure 💯

mtias commented:
Let's try to get the version that is already implemented for the g2-repo here if possible.

miminari commented:
This is nice!
I was just searching for a way to add `object-fit` to core Image Block styles in my own plugins and found this PR.
I would like to also add the object-fit control to Image Block. Have you considered that yet?

ntsekouras commented:
Segmented control landed: https://github.com/WordPress/gutenberg/pull/31937, so I'll refresh this one 😄

ntsekouras committed:
use SegmentedControl

ntsekouras committed:
use useCustomUnits

aristath commented on packages/block-library/src/post-featured-image/dimension-controls.js:
```diff
@@ -0,0 +1,108 @@
+	setAttributes,
+} ) => {
+	const units = useCustomUnits( {
+		availableUnits: [ 'px', '%', 'vw', 'em', 'rem' ],
```
Should this inherit user options from `spacing.units`? example: https://github.com/WordPress/gutenberg/blob/8d5a38a6d708a0a5a1a4ed5afc81f2df9528d6e7/packages/block-editor/src/components/letter-spacing-control/index.js#L25

ntsekouras replied:
I'm not really sure about this, but I've updated it as you have worked more on these. @jorgefilipecosta or @jasmussen  could chime in as well..

ntsekouras committed:
use useSetting

ntsekouras committed:
fix default units

kjellr commented:
This is really helpful! It's working well in my testing:

![featured-image](https://user-images.githubusercontent.com/1202812/126339932-fb752f6b-0d25-407e-8772-76caaa12bddf.gif)

The only thing I'm missing is some sort of reset option: there's no button, and deleting entered values doesn't seem to work either.

ntsekouras committed:
fix resetting

ntsekouras commented:
>The only thing I'm missing is some sort of reset option: there's no button, and deleting entered values doesn't seem to work either.

Good catch! Thanks! Fixed..

kjellr reviewed (approved):
Thanks for fixing the reset. This works well in my testing, and seems like a good v1 in case we want to iterate further later on. 👍

antonyjsmith commented:
This is great, I love it, but this now differs when compared to images blocks yet they are in most respects doing the same thing (full disclosure I think the image block is poor) back to my main point though, this lack of consistency confuses users.

paaljoachim commented:
I am wondering where we can extend these new controls to other blocks where it is natural to add them?
As @antonyjsmith says having consistency is important.
Perhaps something to look into for the Image block? @Mamaduka

antonyjsmith commented:
Consistency is absolutely critical. Decisions needs to be considered prior to implementation and to meet a predefined specification otherwise you end up with what Gutenberg currently is. An inconsistent perpetual beta.

ntsekouras commented:
>Consistency is absolutely critical

I agree 100% here and improvements are always being implemented. This is a big and difficult task and we have to tackle this with discussion, new issues and iterations. While some blocks have similarities, the consolidation of them is not so simple.

There are already some issues about consistency for UX and API [like this one](https://github.com/WordPress/gutenberg/issues/33391) and we will get there step by step 😄 .

NicoHood commented:
Can anybody tell me in which version of Gutenberg this feature has landed? It is not available in wordpress 5.8, even though it was mentioned in the PR description, is that correct?

ntsekouras commented:
👋  @NicoHood !

You can see in every merged PR the Gutenberg version in the `Milestone` section, which in this case is `11.3`. This PR landed after WP 5.8 and will be in the upcoming 5.9 version.

FYI: you can [see here](https://developer.wordpress.org/block-editor/contributors/versions-in-wordpress/) the GB versions in each WP version 😄

NicoHood commented:
Nice, thanks!

Is there also an option to select a WordPress predefined image format (like Middle, Large, Thumbnail etc.)? If not, is this issue being tracked somewhere?

Sorry for being a bit lost here.

ntsekouras commented:
It's tracked here: https://github.com/WordPress/gutenberg/issues/33789

NicoHood commented:
Is there a way to get this feature with the current WordPress 5.8 and maybe a hidden hook? I would like to set all feature images of the query loop to a specific size. I am not sure if using the additional gutenberg plugin is a good idea for a stable running site.

ntsekouras commented:
>Is there a way to get this feature with the current WordPress 5.8 and maybe a hidden hook?

I don't think so. It's in the plugin

markhowellsmead commented:
There's no way to select which registered image size is to be used for the base of this block; e.g. in the post query loop, we need to be able to choose that e.g. `thumbnail` should be used (so that no massive images are loaded). See also https://github.com/WordPress/gutenberg/issues/33789#issuecomment-960159300

ntsekouras commented:
>See also #33789 (comment)

That's the existing issue for your request. Thanks for commenting there.

markhowellsmead commented:
Have already done so: https://github.com/WordPress/gutenberg/issues/33789#issuecomment-961730413
